docs: Add GitHub instructions (#24)
codacy/pulse-user-docs: docs/one-click-integrations.md

1.  On Pulse, click **Integrations** and select **PagerDuty**.

>     ![PagerDuty integration](images/pagerduty.png)

Reviewer: Nice catch! :+1:

1.  Fill in the necessary details:

    - **PagerDuty subdomain:** Your custom PagerDuty subdomain. For example, `mysubdomain` if you use the URL `https://mysubdomain.pagerduty.com` to access PagerDuty.
    - **Pulse system name:** Name of the [system](https://docs.pulse.codacy.com/#before-you-begin) to associate with the incidents reported by this integration.

1.  Follow the instructions on the Pulse UI to create a new webhook on PagerDuty using the provided URL. For example:

    ![PagerDuty webhook](images/pagerduty-webhook.png)

## GitHub

Pulse integrates directly with GitHub to receive data about deployments and changes.

To set up the GitHub integration:

1.  On Pulse, click **Integrations** and select **GitHub**.

    ![GitHub integration](images/ghi-setup.png)

> 1.  Click **Install GitHub App**
> 
> 1.  Follow the instructions on the GitHub UI to install the app in the desired organization.

Reviewer: Here we could condense the two actions in a single step since the process is quite straightforward:

```suggestion
1.  Click **Install GitHub App** and follow the instructions on the GitHub UI to install the app in the desired organization.
```

    ![GitHub webhook](images/ghi-github-install.png)

1.  Wait while Pulse creates the integration and the webhook on GitHub for you.

    ![GitHub webhook](images/ghi-creating.png)

1.  In case of success you should be redirected to a page like:

    ![GitHub webhook](images/ghi-ok.png)

> 1.  If there is an error as displayed bellow please [contact support](mailto:pulsesupport@codacy.com).

Reviewer: Small typo:

```suggestion
1.  If there is an error as displayed below please [contact support](mailto:pulsesupport@codacy.com).
```

> 1.  If there is an error as displayed bellow please [contact support](mailto:pulsesupport@codacy.com).

Reviewer 2: ```suggestion
1.  If there is an error as displayed below please [contact support](mailto:pulsesupport@codacy.com).
```

TIL:
![image](https://user-images.githubusercontent.com/11428771/105383230-cc2e1d00-5c08-11eb-853a-ebaaa65fa648.png)

Reviewer: I see this typo a lot, most probably because the spell-checkers don't signal the word since it **does** exist. :slightly_smiling_face:

Reviewer 2: That's exactly what I was thinking, because I'm pretty sure @rtfpessoa has a spell-checker on VSCode

> 1.  In case of success you should be redirected to a page like:
> 
>     ![GitHub webhook](images/ghi-ok.png)
> 
> 1.  If there is an error as displayed bellow please [contact support](mailto:pulsesupport@codacy.com).
> 
>     ![GitHub webhook](images/ghi-error.png)

Reviewer: I would leave the results of the action outside the numbered list since these are no longer actionable steps that the user must perform:

```suggestion
In case of success you should be redirected to a page like:

![GitHub webhook](images/ghi-ok.png)

If there is an error as displayed below please [contact support](mailto:pulsesupport@codacy.com).

![GitHub webhook](images/ghi-error.png)
```

    - **Pulse system name:** Name of the [system](https://docs.pulse.codacy.com/#before-you-begin) to associate with the incidents reported by this integration.

1.  Follow the instructions on the Pulse UI to create a new webhook on PagerDuty using the provided URL. For example:

    ![PagerDuty webhook](images/pagerduty-webhook.png)

## GitHub

Pulse integrates directly with GitHub to receive data about deployments and changes.

To set up the GitHub integration:

1.  On Pulse, click **Integrations** and select **GitHub**.

    ![GitHub integration](images/ghi-setup.png)

1.  Click **Install GitHub App** and follow the instructions on the GitHub UI to install the app in the desired organization.

    ![GitHub webhook](images/ghi-github-install.png)

1.  Wait while Pulse creates the integration and the webhook on GitHub for you.

    ![GitHub webhook](images/ghi-creating.png)

In case of success you should be redirected to a page like:

![GitHub webhook](images/ghi-ok.png)

If there is an error as displayed below please [contact support](mailto:pulsesupport@codacy.com).

![GitHub webhook](images/ghi-error.png)

> ### How it works

Reviewer: Since these details about how the feature works may impact the decision of the users to use the integration or not, I would place this part right before the step by step instructions on how to activate the integration.

For example, using an "important" admonition card like [the one used here](https://docs.codacy.com/repositories-configure/code-patterns/#import-patterns).

Author: Added the change bellow, what do you think?

Reviewer: Looks good, I committed your suggestion.

> - We rely on tags to collect deployments by selecting the ones that are [semver](https://semver.org) non pre-release

Reviewer: Suggestion to clarify how we use tags to obtain information about deployments:

```suggestion
- We rely on tags to collect deployments by selecting non pre-release versions that follow the [SemVer](https://semver.org) convention
```

- For [annotated tags](https://git-scm.com/book/en/v2/Git-Basics-Tagging#_annotated_tags), we know the creation date of the tag itself,
  while for [lightweight tags](https://git-scm.com/book/en/v2/Git-Basics-Tagging#_lightweight_tags), there is no creation date,
  so we rely on the timestamp when our webhook is invoked.
  Since webhook calls can be delayed, the deployment timestamp on Pulse might be imprecise, impacting the `Lead time for changes` metric.

> - We use the commits between two tags to know what changes(commits) existed

Reviewer: Add missing space and suggestion to clarify sentence:

```suggestion
- We use the commits between two tags to know what changes (commits) belong to a deployment
```

> - We use the commits between two tags to know what changes(commits) existed

Author: ```suggestion
!!! important
    **How it works:**
    - We rely on tags to collect deployments by selecting the ones that are [semver](https://semver.org) non pre-release
    - For [annotated tags](https://git-scm.com/book/en/v2/Git-Basics-Tagging#_annotated_tags), we know the creation date of the tag itself,
      while for [lightweight tags](https://git-scm.com/book/en/v2/Git-Basics-Tagging#_lightweight_tags), there is no creation date,
      so we rely on the timestamp when our webhook is invoked.
      Since webhook calls can be delayed, the deployment timestamp on Pulse might be imprecise, impacting the `Lead time for changes` metric.
    - We use the commits between two tags to know what changes(commits) existed
```

> ### How it works
> 
> - We rely on tags to collect deployments by selecting non pre-release versions that follow the [SemVer](https://semver.org) convention
> 
> - For [annotated tags](https://git-scm.com/book/en/v2/Git-Basics-Tagging#_annotated_tags), we know the creation date of the tag itself,
>   while for [lightweight tags](https://git-scm.com/book/en/v2/Git-Basics-Tagging#_lightweight_tags), there is no creation date,
>   so we rely on the timestamp when our webhook is invoked.
>   Since webhook calls can be delayed, the deployment timestamp on Pulse might be imprecise, impacting the `Lead time for changes` metric.
> 
> - We use the commits between two tags to know what changes (commits) belong to a deployment

Author: ```suggestion
!!! important
    **How it works:**
    - We rely on tags to collect deployments by selecting non pre-release versions that follow the [SemVer](https://semver.org) convention
    - For [annotated tags](https://git-scm.com/book/en/v2/Git-Basics-Tagging#_annotated_tags), we know the creation date of the tag itself,
      while for [lightweight tags](https://git-scm.com/book/en/v2/Git-Basics-Tagging#_lightweight_tags), there is no creation date,
      so we rely on the timestamp when our webhook is invoked.
      Since webhook calls can be delayed, the deployment timestamp on Pulse might be imprecise, impacting the `Lead time for changes` metric.
    - We use the commits between two tags to know what changes (commits) belong to a deployment
```

    ![PagerDuty webhook](images/pagerduty-webhook.png)

## GitHub

Pulse integrates directly with GitHub to receive data about deployments and changes.

!!! important
    Consider the following before using the GitHub integration:

    -   The integration obtains information about deployments from Git tags that follow the [SemVer](https://semver.org) convention, excluding pre-release versions.

    -   The integration uses the creation date of [annotated tags](https://git-scm.com/book/en/v2/Git-Basics-Tagging#_annotated_tags) as the deployment date.
    
        However, since [lightweight tags](https://git-scm.com/book/en/v2/Git-Basics-Tagging#_lightweight_tags) don't have this information, in this case the deployment date is the timestamp when the integration receives the webhook call. Since webhook calls can be delayed, the deployment date on Pulse might be imprecise, impacting the metric **Lead time for changes**.

    -   The integration obtains the set of changes that belong to a deployment from the list of commits between the tag of that deployment and the previous tag.
    

>         Because of this, the integration doesn't consider the first SemVer tag in the repository as a deployment because it's not possible to calculate the list of changes for that deployment.

Reviewer: @afsmeira and @rtfpessoa, I added the [detail about the integration "discarding" the first SemVer tag](https://codacy.atlassian.net/browse/PUL-227?focusedCommentId=45157).

Author: possible is a strong word. We could simply consider all the commits before that tag. I guess this was more of a decision.
Maybe @DReigada can help with the wording.

Reviewer: You're right. Perhaps:

> Because of this, the integration discards the first SemVer tag in the repository because there isn't a sure way to calculate the list of changes for that deployment.

Reviewer 2: I think it would be nice to refer the "previous tag", since that is the actual problem

> Because of this, the integration discards the first SemVer tag in the repository since there is no previous tag to compare with.

1.  On Pulse, click **Integrations** and select **PagerDuty**.

    ![PagerDuty integration](images/pagerduty.png)

1.  Fill in the necessary details:

    - **PagerDuty subdomain:** Your custom PagerDuty subdomain. For example, `mysubdomain` if you use the URL `https://mysubdomain.pagerduty.com` to access PagerDuty.
    - **Pulse system name:** Name of the [system](https://docs.pulse.codacy.com/#before-you-begin) to associate with the incidents reported by this integration.

1.  Follow the instructions on the Pulse UI to create a new webhook on PagerDuty using the provided URL. For example:

    ![PagerDuty webhook](images/pagerduty-webhook.png)

## GitHub

Pulse integrates directly with GitHub to receive data about deployments and changes.

!!! important
    Consider the following before using the GitHub integration:

>     -   The integration obtains information about deployments from Git tags that follow the [SemVer](https://semver.org) convention, excluding pre-release versions.

Reviewer: Out of curiosity, do we also parse versions with the `vX.Y.Z` syntax? [I saw this](https://semver.org/#is-v123-a-semantic-version) on the SemVer specification and thought that there is a chance that some customers are using this alternative syntax.

Author: Not, we do not. But if this becomes a problem we can do it.

    -   The integration obtains the deployment date from either the creation date of [annotated tags](https://git-scm.com/book/en/v2/Git-Basics-Tagging#_annotated_tags) or the timestamp when the integration receives the webhook calls for [lightweight tags](https://git-scm.com/book/en/v2/Git-Basics-Tagging#_lightweight_tags).
    
        However, since webhook calls can be delayed, in this case the deployment date on Pulse could be imprecise and impact the metric **Lead time for changes**.

    -   The integration obtains the set of changes that belong to a deployment from the list of commits between the tag of that deployment and the previous tag.
    
        Because of this, the integration doesn't consider the first SemVer tag in the repository as a deployment because it's not possible to calculate the list of changes for that deployment.

To set up the GitHub integration:

1.  Make sure that you're creating a Git tag on your repositories for each sucessful deployment to production, or whenever you make a new release available to any user of your application:

    ```bash
    git tag -a X.Y.Z -m "<Deployment or release message>"
    ```

>     Where X.Y.Z must be a valid SemVer version without pre-release information.

Reviewer: > > Do you think users could consider changing the way they're using tags to make it possible for the integration to extract all the needed information?
> > I'm thinking if it would be worthwhile to write a sentence or two pointing users in the right direction towards implementing the tag semantics that we're expecting.
> 
> I guess we could add an example, but I am not sure what we could do more. cc\ @DReigada @afsmeira

I made this proposal to include this important bit of information. :+1: 

This first step can be somewhat intimidating, especially if you have lots of repositories and aren't tagging your versions yet. However, since this is essential for the GtiHub integration to work (or at least to be able to calculate all the metrics), I prefer to have this explicit as the very first step of the instructions.
